core/src/main/java/org/elasticsearch/index/query/TermsQueryBuilder.java

```diff
@@ -207,4 +289,130 @@ public void doXContent(XContentBuilder builder, Params params) throws IOExceptio
+            if (fieldType != null) {
+                query = fieldType.termsQuery(terms, context);
+            } else {
+                BytesRef[] filterValues = new BytesRef[terms.size()];
+                for (int i = 0; i < filterValues.length; i++) {
+                    filterValues[i] = BytesRefs.toBytesRef(terms.get(i));
+                }
+                query = new TermsQuery(indexFieldName, filterValues);
+            }
+        } else {
+            BooleanQuery bq = new BooleanQuery(disableCoord);
+            for (Object term : terms) {
+                if (fieldType != null) {
+                    bq.add(fieldType.termQuery(term, context), BooleanClause.Occur.SHOULD);
+                } else {
+                    bq.add(new TermQuery(new Term(indexFieldName, BytesRefs.toBytesRef(term))), BooleanClause.Occur.SHOULD);
+                }
+            }
+            query = bq;
+        }
+        return query;
+    }
+
+    @Override
+    public QueryValidationException validate() {
+        QueryValidationException validationException = null;
+        if (this.fieldName == null) {
+            validationException = addValidationError("field name cannot be null.", validationException);
+        }
+        if (isTermsLookupQuery()) {
+            validationException = validateTermsLookup(validationException);
+        }
+        return validationException;
+    }
+
+    private QueryValidationException validateTermsLookup(QueryValidationException validationException) {
+        if (lookupType == null) {
+            validationException = addValidationError("[terms] query lookup element requires specifying the type.", validationException);
+        }
+        if (lookupId == null) {
+            validationException = addValidationError("[terms] query lookup element requires specifying the id.", validationException);
+        }
+        if (lookupPath == null) {
+            validationException = addValidationError("[terms] query lookup element requires specifying the path.", validationException);
+        }
+        return validationException;
+    }
+
+    @Override
+    protected TermsQueryBuilder doReadFrom(StreamInput in) throws IOException {
+        TermsQueryBuilder termsQueryBuilder = new TermsQueryBuilder(in.readString(), in.readGenericValue());
+        if (termsQueryBuilder.isTermsLookupQuery()) {
+            termsQueryBuilder.lookupIndex = in.readOptionalString();
+            termsQueryBuilder.lookupType = in.readString();
+            termsQueryBuilder.lookupId = in.readString();
+            termsQueryBuilder.lookupPath = in.readString();
+            termsQueryBuilder.lookupRouting = in.readOptionalString();
+        }
+        return termsQueryBuilder;
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeString(fieldName);
+        out.writeGenericValue(values);
+        if (isTermsLookupQuery()) {
```

Comment:
we should move to have an inner terms lookup builder. BTW You don't necessarily need to have a flag in your class, you just have to serialize what the method returns and read based on that on the other side. If you read correctly the method will return the right result. The flag would anyway depend on other instance members as far as I can see.